### PR TITLE
add APIs for service account volume projection

### PR DESCRIPTION
**Pull request by mikedanese**
ref https://github.com/kubernetes/kubernetes/issues/58790

designed in https://github.com/kubernetes/community/pull/1973

Release note will be included in the implementation.
```release-note
NONE
```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -247,6 +247,12 @@ const (
 	// Implement TokenRequest endpoint on service account resources.
 	TokenRequest utilfeature.Feature = "TokenRequest"
 
+	// owner: @mikedanese
+	// alpha: v1.11
+	//
+	// Implement TokenRequestProjection endpoint on service account resources.
```
Comment (liggitt) [inline pkg/features/kube_features.go]: copy pasta

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -1466,6 +1466,29 @@ type ConfigMapProjection struct {
...
+// volume. This projection can be used to insert a service account token into
+// the pods runtime filesystem for use against APIs (Kubernetes API Server or
+// otherwise).
+type ServiceAccountTokenProjection struct {
+	// Audience is the intended audience of the token. A recipient of a token
+	// must identify itself with an identifier specified in the audience of the
+	// token, and otherwise should reject the token. The audience defaults to the
+	// identifier of the apiserver.
+	//+optional
+	Audience string `json:"audience,omitempty" protobuf:"bytes,1,rep,name=audiences"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go]: s/audiences/audience/

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -1039,6 +1039,15 @@ func validateProjectionSources(projection *core.ProjectedVolumeSource, projectio
 				}
 			}
 		}
+		if projPath := fldPath.Child("serviceAccountToken"); source.ServiceAccountToken != nil {
+			numSources++
+			if !utilfeature.DefaultFeatureGate.Enabled(features.TokenRequestProjection) {
+				allErrs = append(allErrs, field.Forbidden(projPath, "TokenRequestProjection feature is not enabled"))
+			}
+			if source.ServiceAccountToken.ExpirationSeconds < 10*60 {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: should limit maximum to integer values that round-trip reliably in json (e.g. 2^53 or less)

Comment (mikedanese) [reply pkg/apis/core/validation/validation.go]: I limited it to 2^32, which is about 2000 years.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (mikedanese) [reply staging/src/k8s.io/api/core/v1/types.go]: Done.

File: pkg/features/kube_features.go (hunk 1 above)
Comment (mikedanese) [reply pkg/features/kube_features.go]: Done.

**Comment by mikedanese** [comment]
@liggitt @tallclair PTAL

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -1391,6 +1391,27 @@ type ConfigMapProjection struct {
...
+	// identifier of the apiserver.
+	Audience string
+	// ExpirationSeconds is the requested duration of validity of the service
+	// account token. As the token approaches expiration, the kubelet volume
+	// plugin will proactively rotate the service account token. The kubelet will
+	// start trying to rotate the token if the token is older than 80 percent of
+	// its time to live or if the token is older than 24 hours.Defaults to 1 hour
+	// and must be at least 10 minutes.
+	ExpirationSeconds int64
+	// Path is the relative path of the file to project the token into.
```
Comment (tallclair) [inline pkg/apis/core/types.go]: Should there be a default?

Comment (tallclair) [inline pkg/apis/core/types.go]: Relative to the mount point?

File: pkg/apis/core/validation/validation.go (hunk 3 above)
Comment (tallclair) [reply pkg/apis/core/validation/validation.go]: > In the year 4027, humanity is on the brink of extinction. The last remaining humans have been forced into hiding under the onslaught of _kubeflow_. Then one day the attacks stop, the machines go quiet. **overlord-service** didn't properly handle cache invalidation of its service account token nor did it implement smart health checks.

```
expirationSeconds: 4294967296  # This should never expire
```

File: staging/src/k8s.io/api/core/v1/types.go [hunk 5]
```diff
@@ -1466,6 +1466,29 @@ type ConfigMapProjection struct {
...
+	//+optional
+	Audience string `json:"audience,omitempty" protobuf:"bytes,1,rep,name=audience"`
+	// ExpirationSeconds is the requested duration of validity of the service
+	// account token. As the token approaches expiration, the kubelet volume
+	// plugin will proactively rotate the service account token. The kubelet will
+	// start trying to rotate the token if the token is older than 80 percent of
+	// its time to live or if the token is older than 24 hours.Defaults to 1 hour
+	// and must be at least 10 minutes.
+	//+optional
+	ExpirationSeconds int64 `json:"expirationSeconds,omitempty" protobuf:"varint,2,opt,name=expirationSeconds"`
```
Comment (tallclair) [inline staging/src/k8s.io/api/core/v1/types.go]: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required says this should be a pointer

Technically for `Audience` too, but I think we rarely make optional strings pointers... 
```
$ grep -Pzo "(?s)\+optional\n\N*\ \*string\N*\n" types.go | wc -l
14
$ grep -Pzo "(?s)\+optional\n\N*\ string\N*\n" types.go | wc -l
270

$ grep -Pzo "(?s)\+optional\n\N*\ \*int\N*\n" types.go | wc -l
42
$ grep -Pzo "(?s)\+optional\n\N*\ \int\N*\n" types.go | wc -l
38
```

File: staging/src/k8s.io/api/core/v1/types.go [hunk 6]
```diff
@@ -1489,6 +1512,8 @@ type VolumeProjection struct {
 	DownwardAPI *DownwardAPIProjection `json:"downwardAPI,omitempty" protobuf:"bytes,2,opt,name=downwardAPI"`
 	// information about the configMap data to project
 	ConfigMap *ConfigMapProjection `json:"configMap,omitempty" protobuf:"bytes,3,opt,name=configMap"`
+	// information about the serviceAccountToken data to project
```
Comment (tallclair) [inline staging/src/k8s.io/api/core/v1/types.go]: `+optional`? Looks like it's missing from the above lines too.

File: pkg/apis/core/types.go (hunk 4 above)
Comment (liggitt) [reply pkg/apis/core/types.go]: I don't think there should be. the kubelet will be explicit, and I think other consumers should be as well

Comment (mikedanese) [reply pkg/apis/core/types.go]: Add validation to ensure that path is set.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 5 above)
Comment (mikedanese) [reply staging/src/k8s.io/api/core/v1/types.go]: So I asked @liggitt about this on slack (so don't fault his chat style response 😄 ) and I wanted to capture his thoughtful perspective:

>I think there are a few things wrapped up in that recommendation... and the fact that it is so inconsistently followed (both in old and new APIs) makes it very hard to find good patterns
>1. we don't want optional omitempty fields getting serialized. for structs, that requires them to be pointers. "empty" string, bool, and int values get omitted without needing to be pointers
>2. when the "empty" value is a valid value and shouldn't be defaulted to something else, we obviously need to make it a pointer
>3. the case you call out (the empty value is not a valid value and we do want to default it to something else) gets weird
>
>if you don't make it a pointer, and default zero-values, it means we'd accept `{"expirationSeconds":0}` and that would mean `{"expirationSeconds":<default value>}`
when from the outside, you'd expect an error about `"expirationSeconds" must be >= 600`.
> as annoying as pointer scaler values are to deal with, for the sake of sane external behavior, we should probably try to differentiate between "explicitly set foo=0" and "no value set"

So he agrees with you. I'll make this a pointer.

Comment (mikedanese) [reply staging/src/k8s.io/api/core/v1/types.go]: @awly, we talked about this same thing a few days ago so this opinion might be interesting to you as well ^^

**Comment by liggitt** [comment]
/milestone v1.11
/kind feature
/status approved-for-milestone
/priority important-soon
/sig auth
/sig node
/sig storage

remove the generated BUILD files under the pkg/kubelet dir, then lgtm

**Comment by mikedanese** [comment]
PTAL

**Comment by liggitt** [comment]
/lgtm

cc @kubernetes/api-approvers 

adds alpha serviceaccount token volume source as proposed in https://github.com/kubernetes/community/pull/1973

**Pull request merged**